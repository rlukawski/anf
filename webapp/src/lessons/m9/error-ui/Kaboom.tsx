import React, { useEffect } from "react";

import { useErrorScreen } from "./ErrorScreenContext";
import { getUserErrorMessage } from "./messages";

interface KaboomSyncProps {
  shouldFail?: () => boolean;
  notify?(message: string): void
};

const throwAlways = () => true

export const KaboomSync = (props: React.PropsWithChildren<KaboomSyncProps>) => {
  const { shouldFail = throwAlways, notify, children } = props;
  if (shouldFail()) {
    notify?.("KaboomSync: fail");
    throw new Error("Kaboom!");
  }
  notify?.("KaboomSync: pass");
  return <>{children}</>;
};

interface KaboomAsyncProps {
  kaboomFn: () => Promise<unknown>
  layout: "ERROR_PAGE" | "ERROR_MODAL"
}

export const KaboomAsync = (props: KaboomAsyncProps) => {
  const { kaboomFn, layout } = props;

  const { showError } = useErrorScreen();
  useEffect(() => {
    kaboomFn().catch((error: unknown) => {
      showError({
        layout,
        message: getUserErrorMessage(error),
      });
    });
  }, [layout, kaboomFn, showError]);

  return <div>this will blink</div>;
};
