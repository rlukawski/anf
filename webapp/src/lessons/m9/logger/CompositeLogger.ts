import { LogLevel } from "api/logs";
import { Logger, LogParams } from "./Logger.interface";

export class CompositeLogger implements Logger {
  private loggers: Logger[] = [];

  add(logger: Logger) {
    this.loggers.push(logger);
    return this;
  }

  send(level: LogLevel, message: string, params?: LogParams) {
    for (const logger of this.loggers) {
      logger.send(level, message, params);
    }
  }

  debug(message: string, params?: LogParams) {
    for (const logger of this.loggers) {
      logger.debug(message, params);
    }
  }

  info(message: string, params?: LogParams) {
    for (const logger of this.loggers) {
      logger.info(message, params);
    }
  }

  warn(message: string, params?: LogParams) {
    for (const logger of this.loggers) {
      logger.warn(message, params);
    }
  }

  error(message: string, params?: LogParams) {
    for (const logger of this.loggers) {
      logger.error(message, params);
    }
  }
}
