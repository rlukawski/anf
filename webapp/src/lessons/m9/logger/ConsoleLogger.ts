/* eslint-disable no-console */
import { LogLevel } from "api/logs";
import { Logger, LogParams } from "./Logger.interface";

export class ConsoleLogger implements Logger {

  static methodMap = {
    DEBUG: console.debug,
    INFO: console.info,
    WARN: console.warn,
    ERROR: console.error,
  };

  send(level: LogLevel, message: string, params?: LogParams) {
    ConsoleLogger.methodMap[level](`${level}: ${message}`, params || "");
  }

  debug(message: string, params?: LogParams) {
    console.debug(`DEBUG: ${message}`, params || "");
  }

  info(message: string, params?: LogParams) {
    console.info(`INFO: ${message}`, params || "");
  }

  warn(message: string, params?: LogParams) {
    console.warn(`WARN: ${message}`, params || "");
  }

  error(message: string, params?: LogParams) {
    console.error(`ERROR: ${message}`, params || "");
  }
}
