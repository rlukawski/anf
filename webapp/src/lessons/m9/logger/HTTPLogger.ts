import { sendLog, LogLevel } from "api/logs";
import { Logger, LogParams } from "./Logger.interface";

export class HTTPLogger implements Logger {
  private sendLog = async (logLevel: LogLevel, message: string, params?: LogParams) => {
    return sendLog(logLevel, message, params ?? null);
  };

  send = (level: LogLevel, message: string, params?: LogParams) => {
    return this.sendLog(level, message, params);
  };

  debug = (message: string, params?: LogParams) => {
    return this.send("DEBUG", message, params);
  };

  info = (message: string, params?: LogParams) => {
    return this.send("INFO", message, params);
  };

  warn = (message: string, params?: LogParams) => {
    return this.send("WARN", message, params);
  };

  error = (message: string, params?: LogParams) => {
    return this.send("ERROR", message, params);
  };
}
