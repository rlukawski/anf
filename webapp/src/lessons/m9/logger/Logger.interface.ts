import { LogLevel } from "api/logs";

export type LogParams = {
  [index: string]: any;
}

export interface Logger {
  send(level: LogLevel, message: string, params?: LogParams): void;
  debug(message: string, params?: LogParams): void;
  info(message: string, params?: LogParams): void;
  warn(message: string, params?: LogParams): void;
  error(message: string, params?: LogParams): void;
}
