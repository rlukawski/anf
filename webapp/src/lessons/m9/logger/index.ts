export * from './Logger.interface'
export * from './LoggerProvider'
export * from './HTTPLogger'
