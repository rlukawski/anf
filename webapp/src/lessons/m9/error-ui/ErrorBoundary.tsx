import React from "react";

import { ErrorPage } from "./ErrorPage";
import { LoggerContext } from "../logger";
import { getUserErrorMessage } from "./messages";

interface ErrorBoundaryProps extends React.PropsWithChildren {};

type ErrorBoundaryState =
  | {
      hasError: false;
    }
  | {
      hasError: true;
      error: Error;
      errorInfo: React.ErrorInfo;
    };
const initialState: ErrorBoundaryState = {
  hasError: false,
};

export class ErrorBoundary extends React.Component<ErrorBoundaryProps, ErrorBoundaryState> {
  constructor(props: ErrorBoundaryProps) {
    super(props);
    this.state = initialState;
  }

  static contextType = LoggerContext;
  context!: React.ContextType<typeof LoggerContext>; // enforcing the type of this.context

  static getDerivedStateFromError(error: Error) {
    return { hasError: true, error: error };
  }

  formatErrorInfo(errorInfo: React.ErrorInfo) {
    return {
      componentStack: errorInfo.componentStack,
    };
  }

  componentDidCatch(error: Error, errorInfo: React.ErrorInfo) {
    const errorMessage = getUserErrorMessage(error)

    this.context.logger.error(`UI Error Boundary reached: ${errorMessage.type}`, {
      message: error.message,
      ...this.formatErrorInfo(errorInfo),
    });

    this.setState({ errorInfo } as ErrorBoundaryState);
  }

  close() {
    this.setState(initialState);
  }

  render() {
    if (this.state.hasError) {
      const errorMessage = getUserErrorMessage(this.state.error)
      return <ErrorPage errorMessage={errorMessage} />;
    } else {
      return this.props.children;
    }
  }
}
