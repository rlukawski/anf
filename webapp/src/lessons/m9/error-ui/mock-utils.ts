import { ErrorBoundary } from "./ErrorBoundary";

// we want to mock the following since these are side effects we don't want to take part in our tests

// by default, error boundaries display diagnostics information to the console
// since we expect the error in this certain case, we mute it, not to pollute the output
export const mockConsoleError = () => {
  const spy = jest.spyOn(console, "error");
  spy.mockImplementation(() => {});
  return () => {
    spy.mockRestore();
  };
};

// in new versions of react, stack trace includes not only specific code lines but also full file paths
// not only line codes are very likely to break in snapshots, but also different people will have completely
// different full file paths on their local machines
export const mockErrorBoundaryStackTrace = () => {
  const errorInfoSpy = jest
    .spyOn(ErrorBoundary.prototype, "formatErrorInfo")
    .mockImplementationOnce((errorInfo: React.ErrorInfo) => ({
      componentStack: `
      this is a mock component stack
      for snapshot purposes
      don't worry`,
    }));

  return () => {
    errorInfoSpy.mockRestore();
  }
}
