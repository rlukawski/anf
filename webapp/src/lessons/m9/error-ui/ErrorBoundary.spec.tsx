import React from 'react';
import { render } from '@testing-library/react';

import { LoggerProvider } from '../logger';
import { MockLogger } from '../logger/Logger.mock'
import { KaboomSync } from "./Kaboom";
import { ErrorBoundary } from './ErrorBoundary';

import { mockConsoleError, mockErrorBoundaryStackTrace } from './mock-utils';

describe('ErrorBoundary', () => {
  it("should display error message if ErrorBoundary caught an exception", () => {
    const mockLogger = new MockLogger();
    const cleanup = [mockConsoleError(), mockErrorBoundaryStackTrace()]

    const { getByText } = render(
      <LoggerProvider logger={mockLogger}>
        <ErrorBoundary>
          <KaboomSync />
        </ErrorBoundary>
      </LoggerProvider>,
    );

    // sprawdzenie wizualne
    getByText("Oops...");
    getByText("Unexpected error occurred. Please contact Support.");

    // cleanup
    cleanup.forEach(fn => fn())
  });

  it("should send ERROR log if ErrorBoundary caught an exception", () => {
    const mockLogger = new MockLogger();
    const cleanup = [mockConsoleError(), mockErrorBoundaryStackTrace()]

    render(
      <LoggerProvider logger={mockLogger}>
        <ErrorBoundary>
          <KaboomSync />
        </ErrorBoundary>
      </LoggerProvider>
    );

    // sprawdzenie czy błąd został zalogowany
    mockLogger.expectLogs.toMatchSnapshot();

    // cleanup
    cleanup.forEach(fn => fn())
  });
});
