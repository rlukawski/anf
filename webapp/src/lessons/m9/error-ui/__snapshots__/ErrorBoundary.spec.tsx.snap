// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundary should send ERROR log if ErrorBoundary caught an exception 1`] = `
Array [
  Array [
    "ERROR",
    "UI Error Boundary reached: UNEXPECTED_ERROR",
    Object {
      "componentStack": "
      this is a mock component stack
      for snapshot purposes
      don't worry",
      "message": "Kaboom!",
    },
  ],
]
`;
